modules/ding_gatewayf/ding_gatewayf.module

```diff
@@ -0,0 +1,458 @@
+<?php
+/**
+ * @file
+ * Defines hooks into the system and main functionality for the module.
+ */
+
+define('DING_GATEWAYF_LOGIN_URL', 'gatewayf/login');
+define('DING_GATEWAYF_LOGOUT_URL', 'gatewayf/logout');
+define('DING_GATEWAYF_CALLBACK_URL', 'gatewayf/callback');
+
+// Constant used as index into the users session.
+define('DING_GATEWAYF_IS_LOGGED_IN', 'wayf_logged_in');
+define('DING_GATEWAYF_ATTRIBUTES', 'ding_gatewayf_attributes');
+
+/**
+ * Implements hook_permission().
+ */
+function ding_gatewayf_permission() {
+  return array(
+    'configure gatewayf' => array(
+      'title' => t('Configure gateway WAYF'),
+      'description' => t('Allow role to configure gate WAYF settings.'),
+    ),
+  );
+}
+
+/**
+ * Implements hook_menu().
+ */
+function ding_gatewayf_menu() {
+  $items = array();
+
+  // Administration config page.
+  $items['admin/config/ding/gatewayf'] = array(
+    'title' => 'Gateway WAYF',
+    'description' => 'Configure gateway WAYF access (NemID)',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('ding_gatewayf_admin_settings_form'),
+    'access arguments' => array('configure gatewayf'),
+    'file' => 'includes/ding_gatewayf.admin.inc',
+  );
+
+  // Redirect the user to gatewayf for authentication.
+  $items[DING_GATEWAYF_LOGIN_URL] = array(
+    'page callback' => 'ding_gatewayf_login',
+    'access callback' => 'user_is_anonymous',
+    'type' => MENU_CALLBACK,
+  );
+
+  // Redirect the user to gatewayf for logout.
+  $items[DING_GATEWAYF_LOGOUT_URL] = array(
+    'page callback' => 'ding_gatewayf_logout',
+    'access callback' => 'user_is_logged_in',
+    'type' => MENU_CALLBACK,
+  );
+
+  // Gatewayf callback which is limited to anonymous users to prevent extra
+  // login and redirects.
+  $items[DING_GATEWAYF_CALLBACK_URL] = array(
+    'page callback' => 'ding_gatewayf_callback',
+    'access callback' => TRUE,
+    'type' => MENU_CALLBACK,
+  );
+
+  return $items;
+}
+
+/**
+ * Implements hook_theme().
+ */
+function ding_gatewayf_theme($existing, $type, $theme, $path) {
+  return array(
+    'ding_gatewayf_login' => array(
+      'template' => 'ding-gatewayf-login',
+      'variables' => array(
+        'link' => NULL,
+      ),
+      'path' => $path . '/templates',
+    ),
+  );
+}
+
+/**
+ * Implements hook_block_info().
+ */
+function ding_gatewayf_block_info() {
+  return array(
+    'login' => array(
+      'info' => t('Ding GateWAYF login link'),
```

Comment:
I have simply taken the fast way and provided blocks the same way as the previous wayf module ding_wayf do.

If this should be re-done it would require another ticket.